### ISSUE TITLE
Add environments rollouts, memory curator, and SFT JSONL export for agent learning loop

### ISSUE THREAD

**Issue by Dhivya-Bharathy** [MervinPraison/PraisonAI#3330]
# Agent Learning Loop — Environments, Rollouts, Curator, and SFT Export

## Title

Add environments-based rollout evaluation with pass-rate scoring, memory curator, and SFT JSONL export to close the agent quality learning loop

---

## Executive Summary

Agno v2.8.0 ships a **closed-loop learning system** beyond static eval:

| Component | Path | Purpose |
|-----------|------|---------|
| `LearningMachine` | `libs/agno/agno/learn/machine.py` | Orchestrates user profile, session context, entity memory, learned knowledge, decision log |
| `Curator` | `libs/agno/agno/learn/curate.py` | Memory maintenance / pruning |
| `Environment` + `Task` | `libs/agno/agno/environments/environment.py` | Define tasks for repeated agent attempts |
| `run_rollouts()` | `libs/agno/agno/environments/runner.py` | Run agent K times per task, compute pass rates |
| SFT export | `libs/agno/agno/environments/exporters.py` | Export passing attempts to JSONL for fine-tuning |
| AgentOS `/learnings` | `os/routers/learnings/router.py` | REST API for learning data |

PraisonAI has **`EvalSuite`** (`praisonaiagents/eval/suite.py`) with accuracy/performance evaluators and closed [#3104](https://github.com/MervinPraison/PraisonAI/issues/3104) for prompt optimization — but **no environments**, **no multi-attempt rollouts**, **no SFT export pipeline**, and **no LearningMachine-style curator**.

Closed [#967](https://github.com/MervinPraison/PraisonAI/issues/967) created eval framework; [#2312](https://github.com/MervinPraison/PraisonAI/issues/2312) noted unbounded learning stores without telemetry/retention. **No issue** proposes Agno-equivalent environments + SFT export.

This proposal adds **`praisonaiagents/learn/`** and **`praisonaiagents/environments/`** modules integrated with existing eval, memory, and `praisonai train` package.

---

### Who benefits

- **ML/platform teams** improving agents via data flywheel (eval → filter → SFT → redeploy).
- **Product managers** measuring agent reliability with pass-rate metrics per task — not single-shot eval scores.
- **Enterprise customers** requiring demonstrable continuous improvement loops for AI governance.

### Why it matters

Agno's README claims: *"turn your agent platform into a learning loop with simulations and usage data."* PraisonAI markets *"self-improving agents"* but lacks the **infrastructure** to systematically collect, score, and export improvement data at scale.

### ROI

- **Model quality:** 10–30% task pass-rate improvement via SFT on filtered rollouts (industry benchmarks for agent fine-tuning).
- **Cost reduction:** Curator prevents unbounded memory growth (#2312 concern).
- **Competitive:** Matches Agno's `#08_learning` and `environments` cookbooks.

### Competitive advantage

Combines PraisonAI's **`praisonai-train`** package and Terminal-Bench benchmarking with a native learning loop — end-to-end story Agno has but PraisonAI's train package currently lacks orchestration glue.

### Risk of ignoring

"Self-improving agents" remains marketing without measurable loop. Teams serious about agent quality will use Agno environments or build custom harnesses outside PraisonAI.

---

### PraisonAI eval (Verified)

`EvalSuite` in `eval/suite.py`:
- Runs list of evaluators (AccuracyEval, PerformanceEval, etc.)
- Returns `EvalSuiteResult` with overall score
- Single-pass evaluation — not K-attempt rollouts
- No pass-rate statistics across attempts
- No export to training formats

### PraisonAI memory (Verified)

- Rich memory layers (STM, LTM, entity, graph)
- No **`Curator`** for automated pruning/maintenance
- #2312: stores grow unbounded

### PraisonAI train (Verified)

- `praisonai-train` package exists for LLM/vision training
- **Not wired** to eval outcomes or rollout filtering

### Limitations

| Gap | Impact |
|-----|--------|
| No `Environment`/`Task` model | Cannot define benchmark suites declaratively |
| No `run_rollouts()` | No pass@K metrics |
| No SFT export | Manual data wrangling for fine-tuning |
| No decision log | No audit of agent reasoning improvements |
| No curator | Memory cost + quality degradation over time |
| Eval doesn't keep best result | #3104 addressed prompt opt — not rollout selection |

---

### Agno LearningMachine

Coordinates:
- User profile / user memory
- Session context extraction
- Entity memory updates
- Learned knowledge (RAG-backed)
- **Decision log** for audit
- **Curator** invoked on schedule or threshold

### Agno Environments

```python
# Conceptual — from environments/runner.py
results = run_rollouts(
    environment=env,
    agent=agent,
    n_attempts=5,
)
# TaskResult.pass_rate, export passing attempts to JSONL
```

Features:
- Live grid UI for rollout progress (`LiveGrid`)
- Fingerprint-based scoring (`scorer/`)
- Policy fingerprint validation across attempts
- **`exporters.py`** → JSONL for SFT pipelines

### Agno AgentOS integration

- `/learnings` REST router persists learning artifacts
- `/evals` router stores eval runs in DB
- Tracing links rollouts to spans

### Quality assessment

Agno's loop is **production-oriented**: DB persistence, REST APIs, cookbook golden paths in `08_learning/` and `environments/`. PraisonAI's eval is **developer-script oriented** without persistence or training export.

---

### Why this matters

Self-improvement requires **measurement → selection → training → deployment** pipeline. EvalSuite covers measurement only. Without rollouts and export, the loop is open-ended.

### Technical reasoning

PraisonAI already has:
- `eval/` module (extend, don't replace)
- `memory/` module (add curator hooks per #1821 direction)
- `praisonai-train` (add SFT ingest API)
- `trace/protocol.py` (link rollouts to traces)

### Architectural reasoning

```mermaid
flowchart LR
    ENV[Environment Tasks] --> ROLL[run_rollouts]
    ROLL --> SCORE[Scorer / Evaluators]
    SCORE --> FILTER[Filter passing attempts]
    FILTER --> SFT[SFT JSONL Export]
    SFT --> TRAIN[praisonai-train]
    TRAIN --> AGENT[Improved Agent]
    AGENT --> ENV
    MEM[Memory] --> CUR[Curator]
    CUR --> MEM
```

---

### Developer UX

```python
from praisonaiagents.environments import Environment, Task, run_rollouts
from praisonaiagents.environments.exporters import export_sft_jsonl

env = Environment(tasks=[
    Task(id="support-1", input="Reset my password", rubric=...),
])
results = run_rollouts(env, agent=my_agent, n_attempts=5)
export_sft_jsonl(results, path="sft/train.jsonl", min_pass_rate=0.8)
```

### CLI UX

```bash
praisonai eval rollout --env support_tasks.yaml --attempts 5
praisonai eval export-sft --run-id abc123 --output train.jsonl
praisonai learn curator --dry-run  # show what would be pruned
```

### SDK UX

`LearningMachine` class coordinating memory updates post-run with configurable strategies.

### Dashboard UX (Claw)

- Rollout grid visualization (pass/fail glyphs per attempt)
- Pass-rate trends over time
- One-click export to train job

### Learning curve

Medium — new concepts (Environment, rollouts) but familiar to ML engineers.

### Daily workflow

Team runs nightly rollouts in CI; curator runs weekly; train job triggered on pass-rate threshold.

---

### Testing difficulty

Rollouts testable with mock agent returning fixed outputs — no LLM required for harness tests.

### Regression risk

Medium — curator could delete wanted memories; require dry-run default.

### Manual validation

1. Define 3-task environment with deterministic scorer.
2. Run 5 attempts; verify pass_rate math.
3. Export JSONL; verify schema matches `praisonai-train` ingest.

### Edge cases

- All attempts fail → empty export with warning
- Agent timeout mid-rollout → partial TaskResult
- Non-deterministic scorer variance → confidence intervals

---

### Maintenance burden

New modules but isolated from agent hot path — opt-in via `agent.learn=True` or explicit CLI.

### Code ownership

- `praisonaiagents/environments/` — rollout engine
- `praisonaiagents/learn/` — LearningMachine + Curator
- `praisonai-train` — ingest SFT JSONL

### Future scalability

- Distributed rollouts via `praisonai-sandbox` backends
- Persist rollout results to platform DB

---

### Security

- SFT export redacts PII via existing `trace/redact.py` patterns
- Decision log access scoped (ties to JWT scope RBAC on serve — proposed enhancement #2)

### Compliance

- Audit trail of what data entered training sets
- Retention policies on rollout artifacts

### Operations

- Nightly rollout CI job with pass-rate alerts
- Curator prevents disk exhaustion

### Governance

- Admin approval before SFT export includes production sessions

---

### Core types

```python
@dataclass
class Task:
    id: str
    input: str
    expected: Optional[str] = None
    rubric: Optional[Callable[[RunOutput], Score]] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

@dataclass
class Environment:
    name: str
    tasks: List[Task]
    policy_fingerprint: Optional[str] = None

@dataclass
class TaskResult:
    task: Task
    attempts: Tuple[AttemptResult, ...]
    
    @property
    def pass_rate(self) -> float: ...
```

### Rollout runner

```python
def run_rollouts(
    environment: Environment,
    agent: Agent,
    n_attempts: int = 3,
    *,
    max_parallel: int = 1,
    scorer: Optional[Scorer] = None,
) -> RolloutResult: ...
```

Async variant `arun_rollouts()` for serve integration.

### LearningMachine

```python
class LearningMachine:
    def __init__(
        self,
        memory: Memory,
        *,
        enable_user_profile: bool = True,
        enable_decision_log: bool = True,
        curator: Optional[Curator] = None,
    ): ...
    
    def after_run(self, run_output: RunOutput) -> None: ...
```

### Curator

```python
class Curator:
    def prune(self, memory: Memory, *, dry_run: bool = False) -> CuratorReport:
        """Remove stale/contradictory/low-confidence memories per policy."""
```

### SFT exporter

```python
def export_sft_jsonl(
    rollout_result: RolloutResult,
    path: Path,
    *,
    min_score: float = 0.8,
    include_traces: bool = False,
) -> ExportReport: ...
```

JSONL schema documented for `praisonai-train` compatibility.

### Integration with EvalSuite

- `EvalSuite` becomes one scorer backend inside rollouts
- Deprecate nothing — add `EvalSuite.as_rollout_scorer()`

---

## Before vs After Comparison

| Capability | Before | After |
|------------|--------|-------|
| Multi-attempt eval | ❌ | ✅ run_rollouts |
| Pass@K metrics | ❌ | ✅ TaskResult.pass_rate |
| SFT data flywheel | ❌ | ✅ export_sft_jsonl |
| Memory maintenance | ❌ | ✅ Curator |
| Decision audit | ❌ | ✅ decision log |
| Train integration | 🔶 manual | ✅ praisonai-train ingest |

---

### Phase 1 — Environment model + runner (4 weeks)

1. `environments/environment.py`, `runner.py`, `scorer.py`
2. Sync + async rollout execution
3. Unit tests with mock agent

### Phase 2 — SFT export + train integration (2 weeks)

1. `exporters.py` JSONL format
2. `praisonai-train` CLI accepts `--sft-jsonl`

### Phase 3 — LearningMachine + Curator (3 weeks)

1. Post-run hooks updating memory selectively
2. Curator with dry-run + retention policies (addresses #2312)

### Phase 4 — CLI + persistence (2 weeks)

1. `praisonai eval rollout` commands
2. SQLite store for rollout history

### Phase 5 — Claw UI + docs (2 weeks)

1. Rollout visualization
2. Cookbook `examples/learning/` with TEST_LOG

### Migration strategy

Purely additive. Existing EvalSuite scripts unchanged.

### Backward compatibility

100%

---

### Files affected

| Module | Files |
|--------|-------|
| praisonai-agents | `environments/*`, `learn/machine.py`, `learn/curate.py`, `learn/decision_log.py`, `eval/suite.py` (integration) |
| praisonai | `cli/commands/eval.py` (extend), `cli/commands/learn.py` (NEW) |
| praisonai-train | ingest API for SFT JSONL |
| tests | `tests/unit/environments/`, `tests/unit/learn/` |

### Pseudo code — rollout engine

```python
def run_rollouts(environment, agent, n_attempts, scorer=None):
    scorer = scorer or DefaultScorer()
    task_results = []
    for task in environment.tasks:
        attempts = []
        for i in range(n_attempts):
            output = agent.run(task.input)
            score = scorer.score(task, output)
            attempts.append(AttemptResult(output=output, score=score, attempt=i))
        task_results.append(TaskResult(task=task, attempts=tuple(attempts)))
    return RolloutResult(environment=environment, tasks=tuple(task_results))
```

---

### Unit Tests

- Pass rate calculation edge cases (0/5, 5/5)
- Exporter filters below min_score
- Curator dry-run doesn't delete

### Integration Tests

- Rollout with real Agent using mock LLM fixture
- JSONL ingested by praisonai-train parser

### Performance Tests

- 100 tasks × 5 attempts completes within SLA with max_parallel=4

### Security Tests

- PII redaction in exported JSONL

---

## Acceptance Criteria

- [ ] `Environment`, `Task`, `run_rollouts()` exported from `praisonaiagents.environments`
- [ ] `export_sft_jsonl()` produces train-compatible JSONL
- [ ] `LearningMachine` hooks into agent post-run (opt-in)
- [ ] `Curator` with dry-run and retention policy config
- [ ] CLI: `praisonai eval rollout`, `praisonai eval export-sft`
- [ ] 40+ unit tests
- [ ] Documentation: Learning Loop guide with mermaid diagram
- [ ] Example in `examples/learning/rollout_and_export.py`

---

## Expected User Experience

Team defines task suite, runs nightly rollouts, exports passing trajectories, triggers train job, deploys improved agent — measurable self-improvement.

---

## Unexpected User Experience

Curator may surprise users if aggressive defaults — ship conservative defaults + dry-run.

---

## Risks

| Risk | Mitigation |
|------|------------|
| Curator deletes useful memory | Dry-run default; explicit opt-in |
| SFT data quality | Scorer + human review gate |
| Overlap with #3104 eval opt | Document: rollouts = multi-attempt, #3104 = prompt tuning |

---

## Estimated Complexity

**Large** — multiple new modules + train integration.

---

## Priority

**High** — core to "self-improving" brand promise; Agno differentiator; complements existing eval.

---

## References

- Agno: `libs/agno/agno/environments/runner.py`, `learn/machine.py`, `learn/curate.py`, cookbook `08_learning/`, `environments/`
- PraisonAI: `praisonaiagents/eval/suite.py`, `praisonai-train/`, `memory/memory.py`
- Issues: [#967](https://github.com/MervinPraison/PraisonAI/issues/967) closed, [#3104](https://github.com/MervinPraison/PraisonAI/issues/3104) closed, [#2312](https://github.com/MervinPraison/PraisonAI/issues/2312) closed
- Agno docs: https://docs.agno.com/
